kubectl-rook-ceph.sh

```diff
@@ -273,6 +276,97 @@ function run_purge_osd() {
+  echo "$deployment_spec" # for returning the updated deployment_spec
+}  
+
+function update_deployment_spec_image(){
+  deployment_spec=$1
+  alternalte_image=\"$2\"
+  deployment_spec=$(jq .template.spec.containers[0].image="$alternalte_image" <<< $deployment_spec)
+  echo "$deployment_spec" # for returning the updated deployment_spec
+}
+
+function verify_deployment_exist(){
+  if [[ $deployment_name != *"mon"* ]] && [[ $deployment_name != *"osd"* ]]; then
+    fail_error "only mon or osd deployment name can passed for starting debug mode"
+  fi
+  if _="$(KUBECTL_NS_CLUSTER get deployment "$deployment_name")"; then
+    echo "setting debug mode for \"$deployment_name\"" 
+  else  
+    fail_error "deployment $deployment_name doesn't exist"
+  fi
+}
+
+function run_start_debug(){
+  [[ -z "${1:-""}" ]] && fail_error "Missing mon or osd deployment name"
+  deployment_name="$1"
+  alternate_image_flag="${2:-""}"
+  alternate_image="${3:-""}"
+  verify_deployment_exist "$deployment_name"
+
+  # copy the deployment spec before scaling it down
+  deployment_spec=$(KUBECTL_NS_CLUSTER get  deployments  "$deployment_name" -o json | jq -r ".spec")
+  
+  # update the deployment_spec main container to be only a placeholder,
+  # also update image if alternate-image is passed
+  if [ -n "$alternate_image_flag" ]; then
+    if [ "$alternate_image_flag" == "--alternate-image" ]; then
+      if [ "$#" -eq 3 ]; then
+        echo "setting debug image to $alternate_image"
+        deployment_spec=$(update_deployment_spec_image "$deployment_spec" "$alternate_image")
+      else
+        fail_error "if --alternate-image flag is passed then image argument is mandatory"   
+      fi
+    else
+      fail_error "--alternate-image should be passed as flag"
+    fi
+  fi  
+  echo "setting debug command to main container"
+  deployment_spec=$(update_deployment_spec_command "$deployment_spec")
+
+  # scale the deployment to 0
+  KUBECTL_NS_CLUSTER scale deployments "$deployment_name" --replicas=0
+  
+  # create debug deployment
+  cat <<EOF | $TOP_LEVEL_COMMAND create -f -
+    apiVersion: apps/v1
+    kind: Deployment
+    metadata:
+        name: "$deployment_name-debug"
+        namespace: "$ROOK_CLUSTER_NAMESPACE"
+        labels:
+            ceph.rook.io/debug: "true"
+    spec:
+        $deployment_spec
+EOF
+}
+
+function run_stop_debug(){
+  [[ -z "${1:-""}" ]] && fail_error "Missing mon or osd deployment name"
+  deployment_name="$1"
+  if [[ $deployment_name != *"debug" ]]; then
+    deployment_name="$deployment_name-debug"
+  fi
+  verify_deployment_exist "$deployment_name"
+  echo "removing debug mode from \"$deployment_name\""
+
+  # delete the deployment debug pod
+  KUBECTL_NS_CLUSTER delete deployment "$deployment_name"
+  # remove the debug suffix from the deployment
+  deployment_name="${deployment_name%'-debug'}"
+  # scale the deployment to 1
+  KUBECTL_NS_CLUSTER scale deployments "$deployment_name" --replicas=1
+}
```

Comment:
I think the operator should be restarted afterwards so that it re-reconciles the deployments that may have been in debug mode. An OSD or mon may need to be updated, for example.

Reply:
After returning from the debug mode,

The support team might need to do some injections or corrections so the bug can resolve, in between if they reconcile then they might be gone to some different state maybe. 

I would say if it is needed we can support this as an extra flag, thoughts @BlaineEXE ?

Reply:
@travisn might have thoughts as well, but I'm still of the opinion that once debug mode is stopped, the user should have the expectation that the resource should go back to being healthy and up-to-date with the rest of the cluster. I think a restart should be a given here.

Reply:
Deleting a mon or osd deployment may already trigger a reconcile, let's test that first before an operator restart.